src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,210 @@
+        var exportProvider = exportProviderFactory.CreateExportProvider();
+
+        return exportProvider;
+    }
+
+    private static async Task<IExportProviderFactory> GetCompositionConfigurationAsync(
+        ExportProviderCreationArguments args,
+        CancellationToken cancellationToken)
+    {
+        var compositionCacheFile = GetCompositionCacheFilePath(args.CacheDirectory, args.CatalogPrefix, args.AssemblyPaths);
+
+        // Try to load a cached composition.
+        try
+        {
+            if (File.Exists(compositionCacheFile))
+            {
+                args.LogTrace($"Loading cached MEF catalog: {compositionCacheFile}");
+
+                CachedComposition cachedComposition = new();
+                using FileStream cacheStream = new(compositionCacheFile, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 4096, useAsync: true);
+                var exportProviderFactory = await cachedComposition.LoadExportProviderFactoryAsync(cacheStream, args.Resolver, cancellationToken).ConfigureAwait(false);
+
+                return exportProviderFactory;
+            }
+        }
+        catch (Exception ex)
+        {
+            // Log the error, and move on to recover by recreating the MEF composition.
+            args.LogError($"Loading cached MEF composition failed: {ex}");
+        }
+
+        args.LogTrace($"Composing MEF catalog using:{Environment.NewLine}{string.Join($"    {Environment.NewLine}", args.AssemblyPaths)}.");
+
+        var discovery = PartDiscovery.Combine(
+            args.Resolver,
+            new AttributedPartDiscovery(args.Resolver, isNonPublicSupported: true), // "NuGet MEF" attributes (Microsoft.Composition)
+            new AttributedPartDiscoveryV1(args.Resolver));
+
+        var parts = await discovery.CreatePartsAsync(args.AssemblyPaths, progress: null, cancellationToken).ConfigureAwait(false);
+        var catalog = ComposableCatalog.Create(args.Resolver)
+            .AddParts(parts)
+            .WithCompositionService(); // Makes an ICompositionService export available to MEF parts to import
+
+        // Assemble the parts into a valid graph.
+        var config = CompositionConfiguration.Create(catalog);
+
+        // Verify we only have expected errors.
+        ThrowOnUnexpectedErrors(config, catalog, args.ExpectedErrorParts, args.LogError);
+
+        // Try to cache the composition.
+        _cacheWriteTask = WriteCompositionCacheAsync(compositionCacheFile, config, args.PerformCleanup, args.LogError, cancellationToken).ReportNonFatalErrorAsync();
+
+        // Prepare an ExportProvider factory based on this graph.
+        return config.CreateExportProviderFactory();
+    }
+
+    private static string GetCompositionCacheFilePath(string cacheDirectory, string catalogPrefix, ImmutableArray<string> assemblyPaths)
+    {
+        return Path.Combine(cacheDirectory, $"{catalogPrefix}.{ComputeAssemblyHash(assemblyPaths)}{CatalogSuffix}");
```

Comment:
interesting.  so we effectively 'leak' the composition caches over time?  not complainign, just trying to understand.

Reply:
There shouldn't be any leaks of caches, the path includes the VS instance and hive, and any mef compositions under that path that don't match the currently hashed cache file name should be deleted.